### PR TITLE
R4R: Refactor rand module

### PR DESCRIPTION
**Pull request by secret2830**
Resolves: #2092

### CODE REVIEW

File: modules/rand/internal/types/expected_keepers.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authexported "github.com/cosmos/cosmos-sdk/x/auth/exported"
+)
+
+// AccountKeeper defines the expected account keeper used for simulations (noalias)
+type AccountKeeper interface {
+	GetAccount(ctx sdk.Context, addr sdk.AccAddress) authexported.Account
```
Comment (chengwenxi) [inline modules/rand/internal/types/expected_keepers.go]: seems no need

File: modules/rand/internal/types/keys.go [hunk 2]
```diff
@@ -1,7 +1,24 @@
...
+
+const (
+	// ModuleName is the name of the rand module
+	ModuleName = "rand"
+
+	// StoreKey is the string store representation
+	StoreKey = ModuleName
+
+	// TStoreKey is the string transient store representation
+	TStoreKey = "transient_" + ModuleName
```
Comment (chengwenxi) [inline modules/rand/internal/types/keys.go]: seems unused.

File: modules/rand/internal/types/expected_keepers.go (hunk 1 above)
Comment (secret2830) [reply modules/rand/internal/types/expected_keepers.go]: > seems no need

Used for the simulation operations

File: modules/rand/internal/types/keys.go (hunk 2 above)
Comment (secret2830) [reply modules/rand/internal/types/keys.go]: Yeah

File: modules/rand/client/cli/tx.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+	return randTxCmd
+}
+
+// GetCmdRequestRand implements the request-rand command
+func GetCmdRequestRand(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "request-rand",
+		Short:   "Request a random number",
+		Example: "iriscli rand request-rand --block-interval=10",
```
Comment (SegueII) [inline modules/rand/client/cli/tx.go]: iriscli tx rand

File: modules/rand/client/cli/query.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+	return randQueryCmd
+}
+
+// GetCmdQueryRand implements the query-rand command.
+func GetCmdQueryRand(queryRoute string, cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "query-rand",
+		Short:   "Query a random number by the request id",
+		Example: "iriscli rand query-rand --request-id=<request id>",
```
Comment (SegueII) [inline modules/rand/client/cli/query.go]: iriscli query rand

File: modules/rand/client/cli/query.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+	return cmd
+}
+
+// GetCmdQueryRandRequestQueue implements the query-queue command.
+func GetCmdQueryRandRequestQueue(queryRoute string, cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "query-queue",
+		Short:   "Query the random number request queue with an optional height",
+		Example: "iriscli rand query-queue [--gen-height=<generation height>]",
```
Comment (SegueII) [inline modules/rand/client/cli/query.go]: ditto

File: modules/rand/internal/types/genesis.go [hunk 6]
```diff
@@ -4,3 +4,22 @@ package types
...
+func DefaultGenesisState() GenesisState {
+	return GenesisState{
+		PendingRandRequests: map[string][]Request{},
+	}
+}
+
+// ValidateGenesis validates the given rand genesis state
+func ValidateGenesis(data GenesisState) error {
+	return nil
+}
```
Comment (SegueII) [inline modules/rand/internal/types/genesis.go]: validate GenesisState

File: modules/rand/handler.go [hunk 7]
```diff
@@ -1,31 +1,51 @@
...
+		),
+	)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeRequestRand,
+			sdk.NewAttribute(types.AttributeKeyRequestID, hex.EncodeToString(GenerateRequestID(request))),
+			sdk.NewAttribute(types.AttributeKeyGenHeight, fmt.Sprintf("%d", request.Height+int64(msg.BlockInterval))),
+		),
+	)
```
Comment (SegueII) [inline modules/rand/handler.go]: merge two single event into one events
```
ctx.EventManager().EmitEvents(
    sdk.Events{
         sdk.NewEvent{...},
         sdk.NewEvent{...},
    }
)
```

File: modules/rand/abci.go [hunk 8]
```diff
@@ -4,20 +4,21 @@ import (
 	"encoding/hex"
 	"fmt"
 
-	sdk "github.com/irisnet/irishub/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/irisnet/irishub/modules/rand/internal/types"
```
Comment (SegueII) [inline modules/rand/abci.go]: use alias instead of import "internal/types" and "internal/keeper" in "rand" package

File: modules/rand/handler.go [hunk 9]
```diff
@@ -1,31 +1,51 @@
 package rand
 
 import (
-	sdk "github.com/irisnet/irishub/types"
+	"encoding/hex"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/irisnet/irishub/modules/rand/internal/types"
```
Comment (SegueII) [inline modules/rand/handler.go]: ditto

File: modules/rand/internal/keeper/keeper_test.go [hunk 10]
```diff
@@ -3,10 +3,10 @@ package keeper
 import (
 	"testing"
```
Comment (SegueII) [inline modules/rand/internal/keeper/keeper_test.go]: use "github.com/stretchr/testify/suite"

File: modules/rand/internal/types/expected_keepers.go [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authexported "github.com/cosmos/cosmos-sdk/x/auth/exported"
```
Comment (SegueII) [inline modules/rand/internal/types/expected_keepers.go]: use "auth" instead of "authexported"

File: modules/rand/client/cli/query.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/irisnet/irishub/modules/rand"
```
Comment (SegueII) [inline modules/rand/client/cli/query.go]: circle import

File: modules/rand/client/cli/tx.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	"bufio"
+	"strconv"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"
+	"github.com/irisnet/irishub/modules/rand"
```
Comment (SegueII) [inline modules/rand/client/cli/tx.go]: circle import

File: modules/rand/client/rest/query.go [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+import (
+	"encoding/hex"
+	"fmt"
+	"net/http"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/types/rest"
+	"github.com/gorilla/mux"
+	"github.com/irisnet/irishub/modules/rand"
```
Comment (SegueII) [inline modules/rand/client/rest/query.go]: circle import

File: modules/rand/client/rest/tx.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"net/http"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/rest"
+	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"
+	"github.com/gorilla/mux"
+	"github.com/irisnet/irishub/modules/rand"
```
Comment (SegueII) [inline modules/rand/client/rest/tx.go]: circle import

**Comment by secret2830** [comment]
> delete exported/exported.go if useless?

ok

File: modules/rand/internal/types/expected_keepers.go (hunk 11 above)
Comment (secret2830) [reply modules/rand/internal/types/expected_keepers.go]: I think the module name should represent the corresponding types package (or alias).

Comment (SegueII) [reply modules/rand/internal/types/expected_keepers.go]: yes, u are right

**Comment by secret2830** [comment]
> delete `rational.go` and `rational_test.go` if not use.

Done

**Pull request merged**